File: December2021/C++/Binary_Tree_Tilt.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int sum=0;
public:
    int findTilt(TreeNode* root) {
        dfs(root);
        return sum;
        
    }
    int dfs(TreeNode* root)
    {
        if(!root) return 0;
        int left=dfs(root->left);
        int right=dfs(root->right);
        sum+=abs(left-right);
        return root->val+left+right;
    }
};

File: December2021/BinaryTreeTilt.java
// TC : O(n)
// SC : O(logn)
public class Solution {
	private int result = 0;

	public int findTilt(TreeNode root) {
		postOrder(root);
		return result;
	}

	private int postOrder(TreeNode root) {
		if (root == null) return 0;

		int leftSubstreeSum = postOrder(root.left);
		int rightSubstreeSum = postOrder(root.right);

		result += Math.abs(leftSubstreeSum - rightSubstreeSum);

		return leftSubstreeSum + rightSubstreeSum + root.val;
	}
}